snorkel/contrib/fonduer/fonduer/lf_helpers.py

```diff
@@ -0,0 +1,1045 @@
+
+    :param span: The span whose head Cells are being returned
+    :param axis: Which axis {'row', 'col'} to search. If None, then both row and col are searched.
+    :param infer: If True, then if a Cell is empty, use the contents from the first non-empty Cell to the left of it
+    :param attrib: The token attribute type (e.g. words, lemmas, poses)
+    :param n_min: The minimum n of the ngrams that should be returned
+    :param n_max: The maximum n of the ngrams that should be returned
+    :param lower: If True, all ngrams will be returned in lower case
+    :rtype: a _generator_ of ngrams
+    """
+    spans = [span] if isinstance(span, TemporarySpan) else span.get_contexts()
+    axes = [axis] if axis else ['row', 'col']
+    for span in spans:
+        if not span.sentence.cell:
+            return
+        else:
+            for axis in axes:
+                if getattr(span.sentence, _other_axis(axis) + '_start') == 0:
+                    return
+                for phrase in getattr(_get_head_cell(span.sentence.cell, axis, infer=infer), 'phrases', []):
+                    for ngram in tokens_to_ngrams(getattr(phrase, attrib), n_min=n_min, n_max=n_max, lower=lower):
+                        yield ngram
+
+
+def _get_head_cell(root_cell, axis, infer=False):
+    other_axis = 'row' if axis == 'col' else 'col'
+    aligned_cells = _get_aligned_cells(
+        root_cell, axis, direct=True, infer=infer)
+    return sorted(aligned_cells, key=lambda x: getattr(x, other_axis + '_start'))[0] if aligned_cells else []
+
+
+def _get_axis_ngrams(span, axis, direct=True, infer=False, attrib='words', n_min=1, n_max=1, spread=[0, 0], lower=True):
+    for ngram in get_phrase_ngrams(span, attrib=attrib, n_min=n_min, n_max=n_max, lower=lower):
+        yield ngram
+    if (span.sentence.cell is not None):
+        for phrase in _get_aligned_phrases(span.sentence, axis, direct=direct, infer=infer, spread=spread):
+            for ngram in tokens_to_ngrams(getattr(phrase, attrib), n_min=n_min, n_max=n_max, lower=lower):
+                yield ngram
+
+
+def _get_aligned_cells(root_cell, axis, direct=True, infer=False):
+    aligned_cells = [cell for cell in root_cell.table.cells
+                     if is_axis_aligned(root_cell, cell, axis=axis)
+                     and cell != root_cell]
+    return [_infer_cell(cell, _other_axis(axis), direct=direct, infer=infer)
+            for cell in aligned_cells] if infer else aligned_cells
+
+
+def _get_aligned_phrases(root_phrase, axis, direct=True, infer=False, spread=[0, 0]):
+    return [phrase for cell in root_phrase.table.cells if is_axis_aligned(root_phrase, cell, axis=axis, spread=spread)
+            for phrase in _infer_cell(cell, _other_axis(axis), direct, infer).phrases
+            if phrase != root_phrase]
+
+
+# PhantomCell = namedtuple('PhantomCell','phrases')
+# TODO: fix this function and retest
+def _infer_cell(root_cell, axis, direct, infer):
```

Comment:
This function is currently not used. I also don't really understand what it was going to be used for (we may have to ask Braden for that). 

If we decide that it is not needed, we could completely delete this, and consequently remove the `direct=True` parameters from a bunch of the functions above.